service/matching/incoming_service_manager.go

```diff
@@ -0,0 +1,294 @@
+	}
+
+	err := m.persistence.DeleteNexusIncomingService(ctx, &p.DeleteNexusIncomingServiceRequest{
+		LastKnownTableVersion: m.tableVersion,
+		ServiceID:             service.Id,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	m.tableVersion++
+	delete(m.servicesByName, request.Name)
+	m.services = slices.DeleteFunc(m.services, func(entry *persistencepb.NexusIncomingServiceEntry) bool {
+		return entry.Service.Name == request.Name
+	})
+	close(m.tableVersionChanged)
+	m.tableVersionChanged = make(chan struct{})
+
+	return &matchingservice.DeleteNexusIncomingServiceResponse{}, nil
+}
+
+func (m *incomingServiceManager) ListNexusIncomingServices(
+	ctx context.Context,
+	request *matchingservice.ListNexusIncomingServicesRequest,
+) (*matchingservice.ListNexusIncomingServicesResponse, chan struct{}, error) {
+	if !m.hasLoadedServices.Load() {
+		if err := m.loadServices(ctx); err != nil {
+			return nil, nil, fmt.Errorf("error loading nexus incoming services cache: %w", err)
+		}
+	}
+
+	m.RLock()
+	defer m.RUnlock()
+
+	if request.LastKnownTableVersion != 0 && request.LastKnownTableVersion != m.tableVersion {
+		return nil, nil, fmt.Errorf("%w received: %v expected: %v", p.ErrNexusTableVersionConflict, request.LastKnownTableVersion, m.tableVersion)
+	}
+
+	startIdx := 0
+	if request.NextPageToken != nil {
+		nextServiceID := primitives.UUIDString(request.NextPageToken)
+
+		startFound := false
+		startIdx, startFound = slices.BinarySearchFunc(
+			m.services,
+			&persistencepb.NexusIncomingServiceEntry{Id: nextServiceID},
+			func(a *persistencepb.NexusIncomingServiceEntry, b *persistencepb.NexusIncomingServiceEntry) int {
+				return cmp.Compare(a.Id, b.Id)
+			})
+
+		if !startFound {
+			return nil, nil, fmt.Errorf("could not find service indicated by nexus list incoming services next page token: %w", p.ErrNexusIncomingServiceNotFound)
+		}
+	}
+
+	endIdx := startIdx + int(request.PageSize)
+	if endIdx > len(m.services) {
+		endIdx = len(m.services)
+	}
+
+	var nextPageToken []byte
+	if endIdx < len(m.services) {
+		nextPageToken = []byte(m.services[endIdx].Id)
+	}
+
+	resp := &matchingservice.ListNexusIncomingServicesResponse{
+		TableVersion:  m.tableVersion,
+		NextPageToken: nextPageToken,
+		Entries:       m.services[startIdx:endIdx],
+	}
+
+	return resp, m.tableVersionChanged, nil
+}
+
+func (m *incomingServiceManager) loadServices(ctx context.Context) error {
+	m.Lock()
+	defer m.Unlock()
+
+	if m.hasLoadedServices.Load() {
+		// check whether services were loaded while waiting for write lock
+		return nil
+	}
+
+	// reset cached view since we will be paging from the start
+	m.services = []*persistencepb.NexusIncomingServiceEntry{}
+	m.servicesByName = make(map[string]*persistencepb.NexusIncomingServiceEntry)
+
+	finishedPaging := false
+	var pageToken []byte
+
+	for ctx.Err() == nil && !finishedPaging {
+		resp, err := m.persistence.ListNexusIncomingServices(ctx, &p.ListNexusIncomingServicesRequest{
```

Comment:
On FailedPrecondition (table was modified while paging) we should reset everything and restart the loop.